test/relayserver/RelayServer.test.ts

```diff
@@ -0,0 +1,1245 @@
+        contracts: {
+          relayHubAddress: relayHub.address,
+        },
+      });
+    });
+
+    afterEach(function () {
+      config.util.extendDeep(config, originalConfig);
+    });
+
+    it('should initialize relay params (chainId, networkId)', async function () {
+      const [relayOwner] = (await ethers.getSigners()) as [SignerWithAddress];
+
+      const { chainId } = provider.network;
+      const networkId = (await provider.send('net_version', [])) as number;
+
+      const server = getServerInstance({ relayOwner });
+
+      expect(server.chainId).to.not.be.equal(chainId);
+      expect(server.networkId).to.not.be.equal(networkId);
+
+      await server.init();
+
+      expect(server.chainId).to.be.equal(chainId);
+      expect(server.networkId).to.be.equal(networkId);
+    });
+  });
+
+  describe('validation', function () {
+    let relayHub: RelayHub;
+    let relayServer: RelayServer;
+    let relayClient: RelayClient;
+    let recipient: TestRecipient;
+    let hubInfo: HubInfo;
+    let owner: Wallet;
+    let encodedData: string;
+
+    beforeEach(async function () {
+      const fakeDeployVerifier = ethers.Wallet.createRandom();
+      const fakeRelayVerifier = ethers.Wallet.createRandom();
+      relayHub = await deployRelayHub();
+      loadConfiguration({
+        app: basicAppConfig,
+        contracts: {
+          relayHubAddress: relayHub.address,
+          deployVerifierAddress: fakeDeployVerifier.address,
+          relayVerifierAddress: fakeRelayVerifier.address,
+        },
+      });
+      recipient = await deployTestRecipient();
+      const { chainId } = provider.network;
+      const serverUrl = buildServerUrl();
+      setEnvelopingConfig({
+        preferredRelays: [serverUrl],
+        chainId,
+        relayHubAddress: relayHub.address,
+        deployVerifierAddress: fakeDeployVerifier.address,
+        relayVerifierAddress: fakeRelayVerifier.address,
+      });
+      relayClient = new RelayClient();
+      const [relayOwner] = (await ethers.getSigners()) as [SignerWithAddress];
+      relayServer = await getInitiatedServer({ relayOwner });
+      hubInfo = relayServer.getChainInfo();
+      owner = ethers.Wallet.createRandom();
+      AccountManager.getInstance().addAccount(owner);
+      encodedData = recipient.interface.encodeFunctionData('emitMessage', [
+        'hello',
+      ]);
+    });
+
+    afterEach(function () {
+      config.util.extendDeep(config, originalConfig);
+    });
+
+    describe('validateInputTypes', function () {
+      it('should throw on undefined field of request', async function () {
+        const userDefinedRelayRequestBody: UserDefinedRelayRequest = {
+          request: {
+            from: owner.address,
+            to: recipient.address,
+            data: encodedData,
+            tokenContract: constants.AddressZero,
+            nonce: 1,
+          },
+          relayData: {
+            callForwarder: constants.AddressZero,
+          },
+        };
```

Comment:
I think we could store this object at an upper layer and change what's necessary for the test (after deep copying the object)

Reply:
updated